src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/Runners/AppRunner.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.TestCommon;
+using Microsoft.Diagnostics.Monitoring.UnitTests.Options;
+using System;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Diagnostics.Monitoring.UnitTests.Runners
+{
+    /// <summary>
+    /// Runner for running the unit test application.
+    /// </summary>
+    internal sealed class AppRunner : IAsyncDisposable
+    {
+        private readonly LoggingRunnerAdapter _adapter;
+
+        private readonly ITestOutputHelper _outputHelper;
+
+        private readonly TaskCompletionSource<string> _readySource =
+            new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        private readonly DotNetRunner _runner = new();
+
+        private TaskCompletionSource<object> _currentCommandSource;
+
+        private bool _isDiposed;
+
+        /// <summary>
+        /// The path of the currently executing assembly.
+        /// </summary>
+        private static string CurrentExecutingAssemblyPath =>
+            Assembly.GetExecutingAssembly().Location;
+
+        /// <summary>
+        /// The path to the application.
+        /// </summary>
+        private static string AppPath =>
+            CurrentExecutingAssemblyPath
+                .Replace(Assembly.GetExecutingAssembly().GetName().Name, "Microsoft.Diagnostics.Monitoring.UnitTestApp");
```

Comment:
Nit: There's a subtle bug here if you have a subfolder with the same name as the assembly. Consider Path.Combine(Path.GetDirectoryName(...),...)

Reply:
We need to do replace because this binary is at:

`artifacts\bin\Microsoft.Diagnostics.Monitoring.UnitTests\Debug\net5.0\Microsoft.Diagnostics.Monitoring.UnitTests.dll`

which needs to be transformed to:

`artifacts\bin\Microsoft.Diagnostics.Monitoring.UnitTestApp\Debug\net5.0\Microsoft.Diagnostics.Monitoring.UnitTestApp.dll`

So, we are intentionally replacing the folder name as well as the file name. I don't think there will be another folder that coincidentally has the same name, unless we really go out of our way to bin-place into particular folders.

If we did build up the path, we'd have to start at `artifacts\bin`, append the new binary name, append the configuration (which we don't have available in C# without a bunch of #ifdef), append the TFM (which we don't have available in C# without a bunch of #ifdef, and then append the binary name again.

Reply:
Thanks for the detailed explanation. I was under the false assumption that execution would combine these into one place for the test run (i.e. the out folder).

Reply:
> Thanks for the detailed explanation. I was under the false assumption that execution would combine these into one place for the test run (i.e. the out folder).

For .NET Framework, that is the case. For .NET Core, they are executed in-place.